src/miner_val_heartbeat.erl

```diff
@@ -0,0 +1,118 @@
+%%% API
+%%%===================================================================
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+init([]) ->
+    {ok, Address0, SigFun, _ECDHFun} = blockchain_swarm:keys(),
+    Address = libp2p_crypto:pubkey_to_bin(Address0),
+    case blockchain_worker:blockchain() of
+        undefined ->
+            erlang:send_after(500, self(), chain_check),
+            {ok, #state{address = Address,
+                        sigfun = SigFun}};
+        Chain -> 
+            ok = blockchain_event:add_handler(self()),
+            {ok, #state{address = Address,
+                        sigfun = SigFun,
+                        chain = Chain}}
+    end.
+
+handle_call(_Request, _From, State) ->
+    lager:warning("unexpected call ~p from ~p", [_Request, _From]),
+    Reply = ok,
+    {reply, Reply, State}.
+
+handle_cast(_Msg, State) ->
+    lager:warning("unexpected cast ~p", [_Msg]),
+    {noreply, State}.
+
+handle_info({blockchain_event, {add_block, Hash, Sync, _Ledger}},
+            #state{address = Address, sigfun = SigFun} = State) ->
+    Ledger = blockchain:ledger(State#state.chain),
+    case blockchain:config(?election_version, Ledger) of
+        {ok, V} when V >= 5 ->
+            {ok, HBInterval} = blockchain:config(?validator_liveness_interval, Ledger),
+            case blockchain:get_block(Hash, State#state.chain) of
+                {ok, Block} ->
+                    Height = blockchain_block:height(Block),
+                    %% heartbeat server needs to be able to run on an unstaked validator
+                    case blockchain_ledger_v1:get_validator(Address, Ledger) of
+                        {ok, Val} ->
+                            lager:debug("getting validator for address ~p got ~p", [Address, Val]),
+                            case blockchain_ledger_validator_v1:last_heartbeat(Val) of
+                                N when (N + HBInterval) =< Height andalso (not Sync) -> 
+                                    %% we need to construct and submit a heartbeat txn
+                                    {ok, CBMod} = blockchain_ledger_v1:config(?predicate_callback_mod, Ledger),
+                                    {ok, Callback} = blockchain_ledger_v1:config(?predicate_callback_fun, Ledger),
+                                    UnsignedTxn =
+                                        blockchain_txn_validator_heartbeat_v1:new(Address, Height, CBMod:Callback()),
+                                    Txn = blockchain_txn_validator_heartbeat_v1:sign(UnsignedTxn, SigFun),
+                                    lager:debug("submitting txn ~p for val ~p ~p ~p", [Txn, Val, N, HBInterval]),
+                                    blockchain_worker:submit_txn(Txn),
```

Comment:
Just wondering if the fire and forget here for the heartbeat txn is sufficient enough, do we need to do `submit/2`, supply a callback and resend if the heartbeat txn fails say X number of times?

Reply:
This loop might spam a lot of redundant transactions as-is, it might be better to use the callback to wait for the one submitted to clear/fail instead of trying every single block.